agent/pkg/podman/podman.go

```diff
@@ -0,0 +1,481 @@
+	envFile, err := os.OpenFile(
+		envFilePath,
+		os.O_APPEND|os.O_CREATE|os.O_WRONLY,
+		0o600,
+	) // #nosec G304 // We made this filepath, and it is randomized.
+	if err != nil {
+		return nil, fmt.Errorf("creating envfile %s: %w", envFilePath, err)
+	}
+	args = append(args, "--env-file", envFilePath)
+
+	req.ContainerConfig.Env = append(req.ContainerConfig.Env, "DET_NO_FLUENT=true")
+	req.ContainerConfig.Env = append(req.ContainerConfig.Env, "DET_SHIPPER_EMIT_STDOUT_LOGS=False")
+
+	for _, env := range req.ContainerConfig.Env {
+		_, err = envFile.WriteString(env + "\n")
+		if err != nil {
+			return nil, fmt.Errorf("writing to envfile: %w", err)
+		}
+	}
+
+	switch {
+	case req.HostConfig.NetworkMode == bridgeNetworking && s.opts.AllowNetworkCreation:
+		// ?? publish ports only for bridgeNetworking
+		for port := range req.ContainerConfig.ExposedPorts {
+			p := port.Int()
+			args = append(args, "-p", fmt.Sprintf("%d:%d/tcp", p, p))
+		}
+		args = append(args, "--network=bridge")
+	case req.HostConfig.NetworkMode == bridgeNetworking:
+		if err = p.Publish(ctx, docker.NewLogEvent(
+			model.LogLevelDebug,
+			"container requested network virtualization, but network creation isn't allowed; "+
+				"overriding to host networking",
+		)); err != nil {
+			return nil, err
+		}
+		req.HostConfig.NetworkMode = hostNetworking
+		fallthrough
+	case req.HostConfig.NetworkMode == hostNetworking:
+		args = append(args, "--network=host")
+	default:
+		return nil, fmt.Errorf("unsupported network mode %s", req.HostConfig.NetworkMode)
+	}
+
+	archivesPath := filepath.Join(tmpdir, archivesName)
+	mountPoints, wErr := cruntimes.ArchiveMountPoints(ctx, req, p, archivesPath, s.log)
+	if wErr != nil {
+		return nil, fmt.Errorf("determining mount points: %w", err)
+	}
+	for _, m := range mountPoints {
+		args = append(args, "--volume", fmt.Sprintf("%s:%s", path.Join(archivesPath, m), m))
+	}
+
+	for _, m := range req.HostConfig.Mounts {
+		args = hostMountsToPodmanArgs(m, args)
+	}
+
+	// from master task_container_defaults.shm_size_bytes
+	if shmsize := req.HostConfig.ShmSize; shmsize != 4294967296 { // 4294967296 is the default.
+		args = append(args, "--shm-size", fmt.Sprintf("%d", shmsize))
+	}
+
+	args = capabilitiesToPodmanArgs(req, args)
+
+	image := cruntimes.CanonicalizeImage(req.ContainerConfig.Image)
+	args = append(args, image)
+	args = append(args, podmanWrapperEntrypoint)
+	args = append(args, req.ContainerConfig.Cmd...)
+
+	if err = s.pprintPodmanCommand(ctx, args, p); err != nil {
+		return nil, err
+	}
+
+	// #nosec G204 // We launch arbitrary user code as a service.
+	cmd := exec.CommandContext(waitCtx, "podman", args...)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, fmt.Errorf("creating stdout pipe: %w", err)
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return nil, fmt.Errorf("creating stderr pipe: %w", err)
+	}
+	s.wg.Go(func(ctx context.Context) { s.shipPodmanCmdLogs(ctx, stdout, stdcopy.Stdout, p) })
```

Comment:
What does "s.wg.Go()" do?

Reply:
Another question for @stoksc

Reply:
it launches a goroutines, but with the new agent code i wrote, i was very careful to not lose track of goroutines, so `s.wg` is a `errgroupx.Group`, a helper for this i wrote. there are docs, i probably could add more though https://pkg.go.dev/github.com/determined-ai/determined/master/pkg/syncx/errgroupx